### ISSUE TITLE
JsonParser from MismatchedInputException cannot getText() for floating-point value

### ISSUE THREAD

**Issue by joca-bt** [FasterXML/jackson-databind#2770]
I've stumbled upon another situation in which I'm not able to get the textual representation of the current token.

I have the following class:

```
public class Test {
    public Boolean value;
    public Boolean getValue() { return value; }
    public void setValue(Boolean value) { this.value = value; }
}
```

And the following object mapper:

```
ObjectMapper objectMapper = new ObjectMapper();
objectMapper.disable(MapperFeature.ALLOW_COERCION_OF_SCALARS);
```

I'm trying to deserialize something that would result in a `MismatchedInputException`: I'm providing a number to something that expects a boolean:

```
try {
    objectMapper.readValue("{ \"value\": 1.0 }", Test.class);
} catch (MismatchedInputException exception) {
    JsonParser parser = (JsonParser) exception.getProcessor();
    System.out.println(parser.currentToken());
    System.out.println(parser.getText());
} catch (Exception exception) {}
```

This results in:

```
VALUE_NUMBER_FLOAT

```

while I would expect `parser.getText()` to return `"1.0"`.

Related with https://github.com/FasterXML/jackson-databind/issues/2635. However that one was specifically for `InputStream`s, where here it happens for both `String`s and `InputStream`s.

(Jackson 2.11.0)

**Comment by joca-bt** [comment]
I have, I can update it tomorrow when I'm back to work.

**Comment by cowtowncoder** [comment]
Fixed for 2.11.1.

**Comment by joca-bt** [comment]
Sorry for the late reply. Looks like it is only float for booleans.

```
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.MapperFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.exc.MismatchedInputException;
import org.junit.jupiter.api.Test;

public class TestTest {
    @Test
    public void test() {
        readValue("{ \"value\": null }");
        readValue("{ \"value\": 1 }");
        readValue("{ \"value\": 1.0 }");
        readValue("{ \"value\": \"true\" }");
        readValue("{ \"value\": [] }");
        readValue("{ \"value\": {} }");
    }

    private void readValue(String value) {
        try {
            try {
                ObjectMapper objectMapper = new ObjectMapper();
                objectMapper.disable(MapperFeature.ALLOW_COERCION_OF_SCALARS)
                    .setDefaultSetterInfo(JsonSetter.Value.construct(Nulls.FAIL, Nulls.FAIL));
                objectMapper.readValue(value, Data.class);
            } catch (MismatchedInputException exception) {
                JsonParser parser = (JsonParser) exception.getProcessor();
                System.out.println(parser.currentToken());
                System.out.println(parser.getText());
            }
        } catch (Exception exception) {}
    }

    public static class Data {
        public Boolean value;
        public Boolean getValue() { return value; }
        public void setValue(Boolean value) { this.value = value; }
    }
}
```

**Comment by cowtowncoder** [comment]
Thanks, makes sense. Code refactoring for 2.12 had actually resolved that for 2.12 branch, but I fixed it for 2.11.
The general problem is tricky, however: there is no real cure for the fact that parser will get closed, buffers recycled. One thing that could help would be use of `DeserializationProblemHandler`, as that will get called before exception is throw, before parser gets called.
Until then I can fix specific cases to force calling of `parser.getText()` before exception is thrown as that (and only that) will ensure that `String` for `getText()` is aggregated, retained, even if buffers will not be accessible any more.
